ddtrace/contrib/openai/_logging.py

```diff
@@ -0,0 +1,87 @@
+import json
+import threading
+from typing import List
+from typing import TypedDict
+
+from ddtrace.internal.compat import get_connection_response
+from ddtrace.internal.compat import httplib
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.periodic import PeriodicService
+
+
+log = get_logger(__file__)
+
+
+class V2LogEvent(TypedDict):
+    """
+    Note: these attribute names match the corresponding entry in the JSON payload.
+    """
+
+    message: str
+    ddtags: str
+    service: str
+    hostname: str
+    ddsource: str
+    # Additional attributes can be specified on the event
+    # including dd.trace_id and dd.span_id to correlate a trace
+
+
+class V2LogWriter(PeriodicService):
+    """Writer to the Datadog log intake.
+
+    v2/logs:
+        - max payload size: 5MB
+        - max single log: 1MB
+        - max array size 1000
+
+    refs:
+        - https://docs.datadoghq.com/api/v2/logs/#send-logs
+    """
+
+    def __init__(self, site, api_key, interval, timeout):
+        # type: (str, str, float, float) -> None
+        super(V2LogWriter, self).__init__(interval=interval)
+        self._lock = threading.RLock()
+        self._buffer = []  # type: List[V2LogEvent]
+        # match the API limit
+        self._buffer_limit = 1000
+        self._timeout = timeout  # type: float
+        self._api_key = api_key  # type: str
+        self._endpoint = "/api/v2/logs"  # type: str
+        self._site = site  # type: str
+        self._intake = "http-intake.logs.%s" % self._site  # type: str
+        self._headers = {
+            "DD-API-KEY": self._api_key,
+            "Content-Type": "application/json",
+        }
+
+    def enqueue(self, log):
+        # type: (V2LogEvent) -> None
+        with self._lock:
+            if len(self._buffer) >= self._buffer_limit:
+                log.warning("log buffer full (limit is %d), dropping log" % self._buffer_limit)
+                return
+            self._buffer.append(log)
+
+    def on_shutdown(self):
+        # type: (...) -> None
+        self.periodic()
+
+    def periodic(self):
+        # type: () -> None
+        if not self._buffer:
+            return
+
+        num_logs = len(self._buffer)
+        payload = json.dumps(self._buffer)
+        self._buffer = []
+        conn = httplib.HTTPSConnection(self._intake, 443, timeout=self._timeout)
```

Comment:
would be good to add retry

Reply:
I think it's ok to defer. The log samples aren't critical data and the data will still likely be present on the span. Are you good with this @majorgreys?